controllers/registries/replication_controller.go

```diff
@@ -84,27 +83,34 @@ func (r *ReplicationReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	}
 
 	// Fetch the Instance
-	harbor, err := internal.FetchReadyHarborInstance(ctx,
-		replication.Namespace,
-		replication.Spec.ParentInstance.Name,
-		r.Client)
-	if err != nil {
-		if _, ok := err.(internal.ErrInstanceNotFound); ok {
-			helper.PullFinalizer(replication, internal.FinalizerName)
-		} else if _, ok := err.(internal.ErrInstanceNotReady); ok {
-			return ctrl.Result{RequeueAfter: 30 * time.Second}, err
-		} else {
-			replication.Status = v1alpha2.ReplicationStatus{LastTransition: &now}
-		}
+	harbor := v1alpha2.Instance{}
+	harborInstanceName := replication.Spec.ParentInstance.Name
+	harborInstanceNamespace := replication.Namespace
 
-		return r.updateReplicationCR(ctx, nil, originalReplication, replication)
+	instanceExists, err := helper.ObjExists(ctx, r.Client, harborInstanceName, harborInstanceNamespace, &harbor)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if !instanceExists {
+		helper.PullFinalizer(replication, internal.FinalizerName)
+		return ctrl.Result{}, controllererrors.ErrInstanceNotFound(
+			strings.Join([]string{harborInstanceName, harborInstanceNamespace}, "/"))
 	}
```

Comment:
This entire block seems a bit repetitive (at least, I've seen similar code in other controllers). Can we refactor this, maybe by extracting a function?

Reply:
I've tried my best by introducing a new helper function `GetOperationalHarborInstance()`, which gets rid of _some_ but not all duplication in this step. Please re-review :+1: